builtin_kernels/sycl-1.2.1/builtin_kernels.md

```diff
@@ -0,0 +1,83 @@
+# Builtin kernels
+
+| Proposal ID | CP018 |
+|-------------|--------|
+| Name | Builtin kernel support in SYCL |
+| Date of Creation | 12 October 2018 |
+| Target | SYCL 1.2.1 vendor extension |
+| Current Status | _Work In Progress_ |
+| Implemented in | _N/A_ |
+| Reply-to | Ruyman Reyes <ruyman@codeplay.com> |
+| Original author | Ruyman Reyes <ruyman@codeplay.com> |
+
+## Overview
+
+Some OpenCL implementations expose already pre-built and embedded _built-in kernels_ to the users. 
+This functionality is used in some cases to expose optimal implementation of some operations, or to
+access some fixed-function hardware available on the platform as a non-programmable OpenCL kernel.
+
+Nowadays SYCL developers can use the OpenCL interoperability functionality to use built-in kernels whenever required.
+However this can be cumbersome and prone to errors as it forces switching back and forward from OpenCL and SYCL/C++.
+
+This vendor extension to the SYCL 1.2.1 specification exposes an interface to work directly from SYCL with
+OpenCL built-in kernels. 
+
+Builtin-kernels can be created explicitly via a new method in the program class, `create_from_built_in_kernel` and then
+used later on from the kernel object, or can be used implicitly with a explicit dispatch function.
+
+In addition, this vendor extension enables further vendor extensions to create SYCL headers that expose a C++ interface
+to the built-in kernels, enabling compile-time checking of parameters.
+
+## Outside of the scope of this extension
+
+* Host emulation of built-in kernels
+
+## Revisions
+
+This is the first revision of this extension.
+
+## Interface changes
+
+### Program class
+
+The program class is extended with a new method, `create_from_built_in_kernel`, 
+which internally uses the relevant OpenCL call(s) to create a valid program object for the devices associated with the instance of the program.
+Built-in kernels do not require a building stage, so the program state is immediately set to executable.
+
+Built-in kernels are not available on the host-device, and will raise `cl::sycl::invalid_context` exception when calling the method on a host context.
```

Comment:
```suggestion
SYCL runtime may provide a host-device implementation. In the case where no host-device implementation is provided, the runtime will raise `cl::sycl::invalid_context` exception when calling the method on a host context.
```

I think it would be useful to have the possibility to query if a builtin kernel is available on a device, like a `is_built_in_kernel_available`

Reply:
Makes sense, but i'll leave it for the next iteration of the proposal.